### PR TITLE
btcpayserver: optional altcoin support

### PR DESCRIPTION
**Pull request by nixbitcoin**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

In use on https://github.com/fort-nix/nix-bitcoin/pull/369

### CODE REVIEW

**Comment by prusnak** [comment]
@GrahamcOfBorg build btcpayserver

File: pkgs/applications/blockchains/btcpayserver/default.nix [hunk 1]
```diff
@@ -34,8 +36,8 @@ stdenv.mkDerivation rec {
     nuget sources Add -Name tmpsrc -Source $TMP/nuget
     nuget init ${linkFarmFromDrvs "deps" deps} $TMP/nuget
 
-    dotnet restore --source $TMP/nuget BTCPayServer/BTCPayServer.csproj
-    dotnet publish --no-restore --output $out/share/$pname -c Release BTCPayServer/BTCPayServer.csproj
+    dotnet restore --source $TMP/nuget ${optionalString altcoinSupport ''/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false''} BTCPayServer/BTCPayServer.csproj
```
Comment (prusnak) [inline pkgs/applications/blockchains/btcpayserver/default.nix:39]: Why is there a need for `RazorCompileOnBuild=false`? Can you add comment into the source code why is that needed?

File: pkgs/applications/blockchains/btcpayserver/update.sh [hunk 2]
```diff
@@ -8,4 +8,4 @@ echo "Updating nbxplorer"
 ../nbxplorer/update.sh
 echo
 echo "Updating btcpayserver"
-../nbxplorer/util/update-common.sh btcpayserver deps.nix
+../nbxplorer/util/update-common.sh btcpayserver deps.nix '"/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false"'
```
Comment (prusnak) [inline pkgs/applications/blockchains/btcpayserver/update.sh:11]: Same here - why is `RazorCompileOnBuild=false` needed?

**Review by prusnak — CHANGES_REQUESTED** [review]

**Comment by nixbitcoin** [comment]
Very interesting, I'm getting `/build/source/BTCPayServer/BTCPayServer.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation. No packages exist with this id in source(s): tmpsrc` when running with a `deps.nix` generated by `update.sh` without `/p:RazorCompileOnBuild=false`, see for yourself with [this test](https://gist.github.com/nixbitcoin/2886b40c05d79173f0c441385429370c). `/p:RazorCompileOnBuild=false` seems unnecessary in the `default.nix` though, see for yourself with [this test](https://gist.github.com/nixbitcoin/734cd5e23049f7aaa9296bd079bff1c5). Removed there and added comment.

File: pkgs/applications/blockchains/btcpayserver/update.sh (hunk 2 above)
Comment (nixbitcoin) [reply pkgs/applications/blockchains/btcpayserver/update.sh]: Needed, see https://github.com/NixOS/nixpkgs/pull/130162/commits/d17ebc3252fa522ee9e52e61b29501c5f231c32e

File: pkgs/applications/blockchains/btcpayserver/default.nix (hunk 1 above)
Comment (nixbitcoin) [reply pkgs/applications/blockchains/btcpayserver/default.nix]: Actually not needed

**Comment by nixbitcoin** [comment]
> Very interesting, I'm getting `/build/source/BTCPayServer/BTCPayServer.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation. No packages exist with this id in source(s): tmpsrc` when running with a `deps.nix` generated by `update.sh` without `/p:RazorCompileOnBuild=false`, see for yourself with [this test](https://gist.github.com/nixbitcoin/2886b40c05d79173f0c441385429370c).~`/p:RazorCompileOnBuild=false` seems unnecessary in the `default.nix` though, see for yourself with [this test](https://gist.github.com/nixbitcoin/734cd5e23049f7aaa9296bd079bff1c5). Removed there and added comment.~
> 
> EDIT: Needed in both [test](https://gist.github.com/nixbitcoin/2886b40c05d79173f0c441385429370c) failing with `fail: Configuration: This build of BTCPay Server does not support altcoins`

Actually, now I'm not sure. The error might be due to the test script. Running fine with the nix-bitcoin test suite based on https://github.com/nixbitcoin/nixpkgs/commit/50fc1b2b2bcb3438431a2c3c03b6c29e00e940db without `RazorCompileOnBuild=false` in the `default.nix`.

File: pkgs/applications/blockchains/btcpayserver/default.nix [hunk 3]
```diff
@@ -1,7 +1,9 @@
 { lib, stdenv, fetchFromGitHub, fetchurl, linkFarmFromDrvs, makeWrapper,
-  dotnetPackages, dotnetCorePackages
+  dotnetPackages, dotnetCorePackages, altcoinSupport ? false
 }:
 
+with lib;
+
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/btcpayserver/default.nix:6]:
```suggestion
```

File: pkgs/applications/blockchains/btcpayserver/default.nix [hunk 4]
```diff
@@ -34,8 +36,9 @@ stdenv.mkDerivation rec {
     nuget sources Add -Name tmpsrc -Source $TMP/nuget
     nuget init ${linkFarmFromDrvs "deps" deps} $TMP/nuget
 
-    dotnet restore --source $TMP/nuget BTCPayServer/BTCPayServer.csproj
-    dotnet publish --no-restore --output $out/share/$pname -c Release BTCPayServer/BTCPayServer.csproj
+    # RazorCompileOnBuild needed for altcoin support
+    dotnet restore --source $TMP/nuget ${optionalString altcoinSupport ''/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false''} BTCPayServer/BTCPayServer.csproj
+    dotnet publish --no-restore --output $out/share/$pname ${optionalString altcoinSupport "-c Altcoins-Release"} BTCPayServer/BTCPayServer.csproj
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/btcpayserver/default.nix:41]:
```suggestion
    dotnet restore --source $TMP/nuget ${lib.optionalString altcoinSupport ''/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false''} BTCPayServer/BTCPayServer.csproj
    dotnet publish --no-restore --output $out/share/$pname ${lib.optionalString altcoinSupport "-c Altcoins-Release"} BTCPayServer/BTCPayServer.csproj
```

Comment (nixbitcoin) [reply pkgs/applications/blockchains/btcpayserver/default.nix]: Fixed in bf6cac3e7a94de1a5dc0da81899ec489780210a7

File: pkgs/applications/blockchains/btcpayserver/default.nix (hunk 3 above)
Comment (nixbitcoin) [reply pkgs/applications/blockchains/btcpayserver/default.nix]: Fixed in bf6cac3e7a94de1a5dc0da81899ec489780210a7
